File: app/src/main/java/com/remotearthsolutions/expensetracker/services/ETFirebaseMessegingService.java
package com.remotearthsolutions.expensetracker.services;

import android.app.Activity;
import android.app.AlertDialog;
import android.app.NotificationManager;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import androidx.core.app.NotificationCompat;
import com.google.firebase.messaging.FirebaseMessagingService;
import com.google.firebase.messaging.RemoteMessage;
import com.remotearthsolutions.expensetracker.R;
import com.remotearthsolutions.expensetracker.activities.ApplicationObject;
import com.remotearthsolutions.expensetracker.activities.MainActivity;
import com.remotearthsolutions.expensetracker.utils.Constants;

public class ETFirebaseMessegingService extends FirebaseMessagingService {

    @Override
    public void onMessageReceived(RemoteMessage remoteMessage) {

        if (((ApplicationObject) getApplicationContext()).isActivityVisible()) {
            Activity activity = ((ApplicationObject) getApplicationContext()).getCurrentActivity();
            activity.runOnUiThread(() -> showAlertDialog(activity, remoteMessage.getNotification().getBody()));
        } else {
            showNotification(remoteMessage);
        }
    }

    private void showAlertDialog(Activity activity, String message) {
        AlertDialog.Builder builder = new AlertDialog.Builder(activity);

        builder.setTitle(getString(R.string.remotearth_notification));
        builder.setMessage(message);
        builder.setIcon(R.mipmap.ic_logo);
        builder.setCancelable(true);
        builder.setNegativeButton(getString(R.string.ok), (dialog, id) -> dialog.cancel());

        AlertDialog alert = builder.create();
        alert.show();
    }

    private void showNotification(RemoteMessage remoteMessage) {
        Intent intent = new Intent(this, MainActivity.class);
        intent.setFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);

        if (remoteMessage.getData().size() > 0) {
            String message = remoteMessage.getData().get(getString(R.string.message));
            Bundle bundle = new Bundle();
            bundle.putString(Constants.KEY_MESSAGE, message);
            intent.putExtras(bundle);
        }

        PendingIntent pendingIntent = PendingIntent.getActivity(this,
                0,
                intent,
                PendingIntent.FLAG_ONE_SHOT);

        NotificationCompat.Builder builder = new NotificationCompat.Builder(this);
        builder.setContentTitle(getString(R.string.remotearth_notification));
        builder.setContentText(remoteMessage.getNotification().getBody());
        builder.setAutoCancel(true);
        builder.setSmallIcon(R.mipmap.ic_logo);
        builder.setContentIntent(pendingIntent);

        NotificationManager manager = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
        manager.notify(0, builder.build());
    }
}

File: app/src/main/java/com/remotearthsolutions/expensetracker/services/ETFirebaseMessegingService.kt
package com.remotearthsolutions.expensetracker.services

import android.content.Intent
import com.google.firebase.messaging.FirebaseMessagingService
import com.google.firebase.messaging.RemoteMessage
import com.remotearthsolutions.expensetracker.BuildConfig
import com.remotearthsolutions.expensetracker.R
import com.remotearthsolutions.expensetracker.activities.ApplicationObject
import com.remotearthsolutions.expensetracker.activities.main.MainActivity
import com.remotearthsolutions.expensetracker.utils.*

class ETFirebaseMessegingService : FirebaseMessagingService() {

    override fun onNewToken(token: String) {
        super.onNewToken(token)
        println(token)
    }

    override fun onMessageReceived(remoteMessage: RemoteMessage) {
        if ((applicationContext as ApplicationObject).isActivityVisible) {
            val activity = (applicationContext as ApplicationObject).currentActivity
            if (remoteMessage.notification != null) {
                if (remoteMessage.data.isNotEmpty()) {
                    val versionCode = remoteMessage.data[Constants.KEY_VERSION_CODE]?.toInt()
                    if (versionCode == BuildConfig.VERSION_CODE || versionCode == 0) {
                        activity!!.runOnUiThread {
                            AlertDialogUtils.show(
                                activity,
                                null,
                                remoteMessage.notification!!.body,
                                activity.getString(R.string.ok),
                                null,
                                null, null
                            )
                        }
                        with(AnalyticsManager) { logEvent("$PN_VIEWED: ${remoteMessage.notification!!.body}") }
                    }
                }
            } else if (remoteMessage.data.isNotEmpty()) {
                val message = remoteMessage.data[Constants.KEY_MESSAGE]
                val versionCode = remoteMessage.data[Constants.KEY_VERSION_CODE]?.toInt()

                if (message == null || versionCode == null) {
                    return
                }

                if (versionCode == BuildConfig.VERSION_CODE || versionCode == 0) {
                    activity!!.runOnUiThread {
                        AlertDialogUtils.show(
                            activity,
                            null,
                            message,
                            activity.getString(R.string.ok),
                            null,
                            null, null
                        )
                    }
                    with(AnalyticsManager) { logEvent("$PN_VIEWED: $message") }
                }
            }
        } else {
            showNotification(remoteMessage)
        }
    }

    private fun showNotification(remoteMessage: RemoteMessage) {
        val intent = Intent(this, MainActivity::class.java)
        if (remoteMessage.data.isNotEmpty()) {
            val message = remoteMessage.data[Constants.KEY_MESSAGE]
            val versionCode = remoteMessage.data[Constants.KEY_VERSION_CODE]?.toInt()
            if (message == null || versionCode == null) {
                return
            }

            intent.putExtra(Constants.KEY_MESSAGE, message)
            if (versionCode == BuildConfig.VERSION_CODE || versionCode == 0) {
                LocalNotificationManager.showNotification(
                    this,
                    getString(R.string.app_name),
                    message,
                    Utils.getPendingIntent(this, intent, 0)
                )
            }
        }
    }
}
